src/classes/CENR_AcademicProgram_TDTM.cls

```diff
@@ -35,69 +35,139 @@
-            //wasn't in the trigger for being two levels up.
-            List<Program_Enrollment__c> pEnrollments = [select Account__c from Program_Enrollment__c 
-                                                    where Account__c != null and ID in :pEnrollIDsToQuery];
-            Map<Id, Program_Enrollment__c> pEnrollmentsMap = new Map<Id, Program_Enrollment__c>(pEnrollments); 
-            
-            for (Integer i = 0; i < newlist.size(); i++) {
-                Course_Enrollment__c enroll = (Course_Enrollment__c)newlist[i];
-                
+
+            if (pEnrollIDsToQuery.size() > 0) {
+                pEnrollments = queryProgramEnrollments(pEnrollIDsToQuery);
+            }
+
+            for (Integer i = 0; i < newCourseEnrollments.size(); i++) {
+                Map<Id, Program_Enrollment__c> pEnrollmentsMap = new Map<Id, Program_Enrollment__c>(pEnrollments);
+                Course_Enrollment__c enroll = (Course_Enrollment__c)newCourseEnrollments[i];
+
+                Boolean isBeforeUpdate;
+
                 //BEFORE INSERT
-                if(triggerAction == TDTM_Runnable.Action.BeforeInsert) {
-                    //set Academic Program if we can get it through Program Enrollment
-                    if(enroll.Program_Enrollment__c != null) {
-                        Program_Enrollment__c pEnroll = pEnrollmentsMap.get(enroll.Program_Enrollment__c);
-                        if(pEnroll != null) //shouldn't really happen... we already queried for this...
-                            enroll.Account__c = pEnroll.Account__c; 
+                if (triggerAction == TDTM_Runnable.Action.BeforeInsert) {
+                    isBeforeUpdate = FALSE;
+                    if (enroll.Program_Enrollment__c != NULL) {
+                        assignAccountAndAffiliation (pEnrollmentsMap, enroll, NULL, affiliationIsEnabled, isBeforeUpdate);
                     }
                 }
-                
+
                 //BEFORE UPDATE
-                if(triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
-                    Course_Enrollment__c oldEnroll = (Course_Enrollment__c)oldList[i];
-                    //set Academic Program if we can get it through Program Enrollment
-                    if(oldEnroll.Program_Enrollment__c == null && enroll.Program_Enrollment__c != null) {
-                        Program_Enrollment__c pEnroll = pEnrollmentsMap.get(enroll.Program_Enrollment__c);
-                        if(pEnroll != null) //shouldn't really happen... we already queried for this...
-                            enroll.Account__c = pEnroll.Account__c;
+                if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+                    isBeforeUpdate = TRUE;
+                    Course_Enrollment__c oldEnroll = (Course_Enrollment__c)oldCourseEnrollments[i];
+                    if (oldEnroll.Program_Enrollment__c == NULL && enroll.Program_Enrollment__c != NULL) {
```

Comment:
@lifewithryan , would it be better to do the oldEnroll.Affiliation__c == NULL check on line 120 so that I don't need to pass the boolean isBeforeUpdate on line 121 and end up doing another if conditional check on line 161?

Reply:
I'll try to look at this over lunch and this afternoon. Tuesday's are my meeting heavy days starting at 11:30 :)